source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,698 @@
+
+    /* make sure the loop is running so we can schedule a last task. */
+    s_run(event_loop);
+
+    /* cancel outstanding tasks */
+    dispatch_async_and_wait_f(dispatch_loop->dispatch_queue, dispatch_loop, s_dispatch_queue_destroy_task);
+
+    AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: Releasing Dispatch Queue.", (void *)event_loop);
+}
+
+static int s_wait_for_stop_completion(struct aws_event_loop *event_loop) {
+    (void)event_loop;
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_try_schedule_new_iteration(struct dispatch_loop_context *loop, uint64_t timestamp);
+
+static int s_run(struct aws_event_loop *event_loop) {
+    struct dispatch_loop *dispatch_loop = event_loop->impl_data;
+
+    s_lock_cross_thread_data(dispatch_loop);
+    if (dispatch_loop->synced_data.suspended) {
+        AWS_LOGF_INFO(AWS_LS_IO_EVENT_LOOP, "id=%p: Starting event-loop thread.", (void *)event_loop);
+        dispatch_resume(dispatch_loop->dispatch_queue);
+        dispatch_loop->synced_data.suspended = false;
+        s_rlock_dispatch_loop_context(dispatch_loop->context);
+        s_lock_service_entries(dispatch_loop->context);
+        s_try_schedule_new_iteration(dispatch_loop->context, 0);
+        s_unlock_service_entries(dispatch_loop->context);
+        s_runlock_dispatch_loop_context(dispatch_loop->context);
+    }
+    s_unlock_cross_thread_data(dispatch_loop);
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_stop(struct aws_event_loop *event_loop) {
+    struct dispatch_loop *dispatch_loop = event_loop->impl_data;
+
+    s_lock_cross_thread_data(dispatch_loop);
+    if (!dispatch_loop->synced_data.suspended) {
+        dispatch_loop->synced_data.suspended = true;
+        AWS_LOGF_INFO(AWS_LS_IO_EVENT_LOOP, "id=%p: Stopping event-loop thread.", (void *)event_loop);
+        /* Suspend will increase the dispatch reference count. It is required to call resume before
+         * releasing the dispatch queue. */
+        dispatch_suspend(dispatch_loop->dispatch_queue);
+    }
+    s_unlock_cross_thread_data(dispatch_loop);
+
+    return AWS_OP_SUCCESS;
+}
+
+/**
+ * The function decides if we should run this iteration.
+ * Returns true if we should execute an iteration, false otherwise
+ *
+ * The function should be wrapped with dispatch_loop->context.lock to retain the dispatch loop while running.
+ */
+static bool begin_iteration(struct scheduled_service_entry *entry) {
+    struct dispatch_loop *dispatch_loop = entry->dispatch_queue_context->io_dispatch_loop;
+
+    if (!dispatch_loop) {
+        return false;
+    }
+    return true;
+}
+
+/**
+ * Clean up the related resource and determine if we should schedule next iteration.
+ * The function should be wrapped with dispatch_loop->context.lock to retain the dispatch loop while running.
+ * */
+static void end_iteration(struct scheduled_service_entry *entry) {
+
+    struct dispatch_loop_context *context = entry->dispatch_queue_context;
```

Comment:
Let's rename this variable to be consistent in its name across various functions for clarity. context -> dispatch_queue_context